lib/multiplexer/web.go

```diff
@@ -0,0 +1,197 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package multiplexer
+
+import (
+	"context"
+	"crypto/tls"
+	"io"
+	"net"
+	"sync/atomic"
+	"time"
+
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/tlsca"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/sirupsen/logrus"
+)
+
+// WebListenerConfig is the web listener configuration.
+type WebListenerConfig struct {
+	// Listener is the listener that accepts tls connections.
+	Listener net.Listener
+	// ReadDeadline is a connection read deadline during the TLS handshake.
+	ReadDeadline time.Duration
+	// Clock is a clock to override in tests.
+	Clock clockwork.Clock
+}
+
+// CheckAndSetDefaults verifies configuration and sets defaults.
+func (c *WebListenerConfig) CheckAndSetDefaults() error {
+	if c.Listener == nil {
+		return trace.BadParameter("missing parameter Listener")
+	}
+	if c.ReadDeadline == 0 {
+		c.ReadDeadline = defaults.HandshakeReadDeadline
+	}
+	if c.Clock == nil {
+		c.Clock = clockwork.NewRealClock()
+	}
+	return nil
+}
+
+// NewWebListener returns a new web listener.
+func NewWebListener(cfg WebListenerConfig) (*WebListener, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	context, cancel := context.WithCancel(context.Background())
+	return &WebListener{
+		log:         logrus.WithField(trace.Component, "mxweb"),
+		cfg:         cfg,
+		webListener: newListener(context, cfg.Listener.Addr()),
+		dbListener:  newListener(context, cfg.Listener.Addr()),
+		cancel:      cancel,
+		context:     context,
+	}, nil
+}
+
+// WebListener multiplexes tls connections between web and database listeners
+// based on the client certificate.
+type WebListener struct {
+	log         logrus.FieldLogger
+	cfg         WebListenerConfig
+	webListener *Listener
+	dbListener  *Listener
+	cancel      context.CancelFunc
+	context     context.Context
+	isClosed    int32
+}
+
+// Web returns web listener.
+func (l *WebListener) Web() net.Listener {
+	return l.webListener
+}
+
+// DB returns database access listener.
+func (l *WebListener) DB() net.Listener {
+	return l.dbListener
+}
+
+// Serve starts accepting and forwarding tls connections to appropriate listeners.
+func (l *WebListener) Serve() error {
+	backoffTimer := time.NewTicker(5 * time.Second)
+	defer backoffTimer.Stop()
+	for {
+		conn, err := l.cfg.Listener.Accept()
+		if err == nil {
+			tlsConn, ok := conn.(*tls.Conn)
+			if !ok {
+				l.log.Errorf("Expected *tls.Conn, got %T.", conn)
+				conn.Close()
+				continue
+			}
+			go l.detectAndForward(tlsConn)
+			continue
+		}
```

Comment:
nit: for readability in my option is more intuitive to handle error case inside if statement